zcash_proofs/src/lib.rs

```diff
@@ -60,56 +78,160 @@ pub fn default_params_folder() -> Option<PathBuf> {
+)]
 pub fn download_parameters() -> Result<(), minreq::Error> {
+    download_sapling_parameters(None).map(|_sapling_paths| ())
+}
+
+/// Download the Zcash Sapling parameters if needed, and store them in the default location.
+/// Always checks the hashes of the files, even if they didn't need to be downloaded.
+///
+/// This mirrors the behaviour of the `fetch-params.sh` script from `zcashd`.
+///
+/// Use `timeout` to set a timeout in seconds for each file download.
+/// If `timeout` is `None`, a timeout can be set using the `MINREQ_TIMEOUT` environmental variable.
+///
+/// Returns the paths to the downloaded files.
+#[cfg(feature = "download-params")]
+#[cfg_attr(docsrs, doc(cfg(feature = "download-params")))]
+pub fn download_sapling_parameters(
+    timeout: Option<u64>,
+) -> Result<SaplingParameterPaths, minreq::Error> {
+    let spend = fetch_params(SAPLING_SPEND_NAME, SAPLING_SPEND_HASH, timeout)?;
+    let output = fetch_params(SAPLING_OUTPUT_NAME, SAPLING_OUTPUT_HASH, timeout)?;
+
+    Ok(SaplingParameterPaths { spend, output })
+}
+
+/// Download the Zcash Sprout parameters if needed, and store them in the default location.
+/// Always checks the hash of the file, even if it didn't need to be downloaded.
+///
+/// This mirrors the behaviour of the `fetch-params.sh` script from `zcashd`.
+///
+/// Use `timeout` to set a timeout in seconds for the file download.
+/// If `timeout` is `None`, a timeout can be set using the `MINREQ_TIMEOUT` environmental variable.
+///
+/// Returns the path to the downloaded file.
+#[cfg(feature = "download-params")]
+#[cfg_attr(docsrs, doc(cfg(feature = "download-params")))]
+pub fn download_sprout_parameters(timeout: Option<u64>) -> Result<PathBuf, minreq::Error> {
+    fetch_params(SPROUT_NAME, SPROUT_HASH, timeout)
+}
+
+/// Download the specified parameters if needed, and store them in the default location.
+/// Always checks the hash of the file, even if it didn't need to be downloaded.
+///
+/// Returns the path to the downloaded file.
+#[cfg(feature = "download-params")]
+#[cfg_attr(docsrs, doc(cfg(feature = "download-params")))]
+fn fetch_params(
+    name: &str,
+    expected_hash: &str,
+    timeout: Option<u64>,
+) -> Result<PathBuf, minreq::Error> {
+    use std::io::BufWriter;
+
     // Ensure that the default Zcash parameters location exists.
     let params_dir = default_params_folder().ok_or_else(|| {
         io::Error::new(io::ErrorKind::Other, "Could not load default params folder")
     })?;
     std::fs::create_dir_all(&params_dir)?;
 
-    let fetch_params = |name: &str, expected_hash: &str| -> Result<(), minreq::Error> {
-        use std::io::Write;
-
-        // Download the parts directly (Sapling parameters are small enough for this).
-        let part_1 = minreq::get(format!("{}/{}.part.1", DOWNLOAD_URL, name)).send()?;
-        let part_2 = minreq::get(format!("{}/{}.part.2", DOWNLOAD_URL, name)).send()?;
```

Comment:
It's necessary for us to host these files in two parts because of [CloudFlare's maximum cached file size limit of 512 MB](https://developers.cloudflare.com/workers/platform/limits#response-limits). 

(It's very important that the files fit in the cache to prevent "denial of wallet" attacks. We are actually on an Enterprise plan at the moment so our *hard* limit is currently 5GB which is why it's safe to keep hosting those single-file deprecated URLs, but after people stopped attacking us it's looking like we might not actually need to pay for Enterprise plan so I don't want to guarantee that the unsplit files will remain.)

Reply:
Thanks, fixed in commit 76f1c15.

Reply:
Thanks!